src/psyclone/psyir/transformations/hoist_loop_bound_expr_trans.py

```diff
@@ -0,0 +1,148 @@
+places them in integer scalar assignments before the loop.
+
+'''
+
+from psyclone.psyir.nodes import Routine, Literal, Reference, \
+    StructureReference, Assignment
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.transformations.loop_trans import LoopTrans, \
+    TransformationError
+
+
+class HoistLoopBoundExprTrans(LoopTrans):
+    '''This transformation moves complex bounds expressions out of the loop
+    construct and places them in integer scalar assignments before the loop.
+
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Loop
+    >>> from psyclone.psyir.transformations import HoistTrans
+    >>> code = ("program test\\n"
+    ...         "  use mymod, only: mytype\\n"
+    ...         "  integer :: i,j,n\\n"
+    ...         "  real :: a(n)\\n"
+    ...         "  do i=mytype%start, UBOUND(a,1)\\n"
+    ...         "    a(i) = 1.0\\n"
+    ...         "  end do\\n"
+    ...         "end program\\n")
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> hoist = HoistLoopBoundExprTrans()
+    >>> hoist.apply(psyir.walk(Loop)[0])
+    >>> print(FortranWriter()(psyir))
+    program test
+      integer :: i
+      integer :: j
+      integer :: n
+      real, dimension(n,n) :: a
+      real :: value
+    <BLANKLINE>
+      value = 1.0
+      do i = 1, n, 1
+        do j = 1, n, 1
+          a(i,j) = value
+        enddo
+      enddo
+    <BLANKLINE>
+    end program test
+    <BLANKLINE>
+
+    '''
+    def apply(self, node, options=None):
+        '''Move complex bounds expressions out of the given loop construct and
+        place them in integer scalar assignments before the loop.
+
+        :param node: target PSyIR loop.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(node, options)
+
+        parent = node.parent
+        position = node.position
+
+        for bound in node.children[0:3]:
+
+            # Skip Literals and non-structure references. Maybe it could be
+            # more selective by allowing arithmetic expressions
+            if isinstance(bound, Literal):
+                continue
+            if isinstance(bound, Reference) and not \
+                    isinstance(bound, StructureReference):
+                continue
+
+            # Create new symbol
+            symbol = node.ancestor(Routine).symbol_table.new_symbol(
+                "loop_bound", symbol_type=DataSymbol, datatype=INTEGER_TYPE
+            )
+            # Move bound expression to an assignment preceding the loop
+            bound.replace_with(Reference(symbol))
+            parent.addchild(Assignment.create(Reference(symbol), bound),
```

Comment:
True, I now forbid the transformation if the parent.parent is a Directive. And I added a TODO since I think this could be handled better if Directives that don't support multiple statements didn't have Schedule as children.